src/main/java/build/buildfarm/common/ScanCount.java

```diff
@@ -0,0 +1,104 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buildfarm.common;
+
+import com.google.common.collect.Sets;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisCluster;
+import redis.clients.jedis.ScanParams;
+import redis.clients.jedis.ScanResult;
+
+///
+/// @class   ScanCount
+/// @brief   Count element of a scan query.
+/// @details Some things are laid out in redis where in order to get their
+///          conceptual size we need to query for wildcard keys and count
+///          non-duplicates. This may not scale because we need to filter non
+///          duplicates.
+///
+public class ScanCount {
+
+  ///
+  /// @brief   Run a scan and count the results.
+  /// @details This is intended to get the size of certain conceptual
+  ///          containers made of many different keys.
+  /// @param   cluster   An established redis cluster.
+  /// @param   query     The query to perform.
+  /// @param   scanCount The count per scan.
+  /// @return  Total number of query results.
+  /// @note    Suggested return identifier: count.
+  ///
+  public static int get(JedisCluster cluster, String query, int scanCount) {
+    Set<String> keys = Sets.newHashSet();
+
+    // JedisCluster only supports SCAN commands with MATCH patterns containing hash-tags.
+    // This prevents us from using the cluster's SCAN to traverse all existing keys.
+    // That's why we choose to scan each of the jedisNode's individually.
+    cluster.getClusterNodes().values().stream()
+        .forEach(
+            pool -> {
+              try (Jedis node = pool.getResource()) {
+                addKeys(cluster, node, keys, query, scanCount);
+              }
+            });
+
+    return keys.size();
+  }
+  ///
+  /// @brief   Scan all entires on node to get key count.
+  /// @details Keys are accumulated onto.
+  /// @param   cluster   An established redis cluster.
+  /// @param   node      A node of the cluster.
+  /// @param   keys      Keys to accumulate.
+  /// @param   query     The query to perform.
+  /// @param   scanCount The count per scan.
+  ///
+  private static void addKeys(
+      JedisCluster cluster, Jedis node, Set<String> keys, String query, int scanCount) {
+    // iterate over all entries via scanning
+    String cursor = "0";
+    do {
+      keys.addAll(scan(node, cursor, query, scanCount));
+
+    } while (!cursor.equals("0"));
```

Comment:
I'm a little confused here about the `cursor` variable in `addKeys(...)` scope. It seems to me `cursor` never changes in this scop and the do-while loop is not necessary.

Reply:
The call to `scan(...,cursor,...)` changes the cursor as an output parameter.  


This is inside the `scan` method:  
```
...
ScanResult scanResult = node.scan(cursor, params);
cursor = scanResult.getCursor(); //<- hopefully not zero until we've reached the end
...
```
---
This code was copied from `WorkerIndexer.java` which does a similar thing.  There may be a cleaner way to write this.  I'd like to hide the cursor details completely and do something that looks more like `ForEach/Map` in both this file and WorkerIndexer.

Reply:
Yeah, i can see `cursor` is intended to be updated inside `scan()`, but `cursor` in `scan()` is just a copy of `addKey()` `cursor` variable. Java just pass it by value.

I did a simple experiment 
```
    private static void stringMutability() {
        String marker = "0";
        changeString(marker);
        System.out.println(marker);
    }

    private static void changeString(String marker) {
        marker = "1";
    }
```
And the printed string is always `0`.

Reply:
To really achieve what you want, `cursor`could be returned by `scan()`, which is not practical here, or could be a member variable of the class so that its state can be seen from a different scope.

Reply:
Oh you're right.  I tested it with too small of data to see the problem.  I'll do what you did in the other diff.

Reply:
fixed